core/src/test/java/kafka/testkit/KafkaClusterTestKit.java

```diff
@@ -404,22 +423,44 @@ private void formatNode(
         boolean writeMetadataDirectory
     ) {
         try {
-            MetaPropertiesEnsemble.Copier copier =
-                new MetaPropertiesEnsemble.Copier(MetaPropertiesEnsemble.EMPTY);
-            for (Entry<String, MetaProperties> entry : ensemble.logDirProps().entrySet()) {
-                String logDir = entry.getKey();
-                if (writeMetadataDirectory || (!ensemble.metadataLogDir().equals(Optional.of(logDir)))) {
-                    log.trace("Adding {} to the list of directories to format.", logDir);
-                    copier.setLogDirProps(logDir, entry.getValue());
+            Formatter formatter = new Formatter();
+            formatter.setNodeId(ensemble.nodeId().getAsInt());
+            formatter.setClusterId(ensemble.clusterId().get());
+            if (writeMetadataDirectory) {
+                formatter.setDirectories(ensemble.logDirProps().keySet());
+            } else {
+                formatter.setDirectories(ensemble.logDirProps().keySet().stream().
+                    filter(d -> !ensemble.metadataLogDir().get().equals(d)).
+                    collect(Collectors.toSet()));
+            }
+            if (formatter.directories().isEmpty()) {
+                return;
+            }
+            formatter.setReleaseVersion(nodes.bootstrapMetadata().metadataVersion());
+            formatter.setFeatureLevel(KRaftVersion.FEATURE_NAME,
+                nodes.bootstrapMetadata().featureLevel(KRaftVersion.FEATURE_NAME));
+            formatter.setUnstableFeatureVersionsEnabled(true);
+            formatter.setIgnoreFormatted(false);
+            formatter.setControllerListenerName("CONTROLLER");
+            if (writeMetadataDirectory) {
+                formatter.setMetadataLogDirectory(ensemble.metadataLogDir().get());
+            } else {
+                formatter.setMetadataLogDirectory(Optional.empty());
+            }
+            if (nodes.bootstrapMetadata().featureLevel(KRaftVersion.FEATURE_NAME) > 0) {
+                StringBuilder dynamicVotersBuilder = new StringBuilder();
+                String prefix = "";
+                for (TestKitNode controllerNode : nodes.controllerNodes().values()) {
+                    int port = socketFactoryManager.
+                        getOrCreatePortForListener(controllerNode.id(), "CONTROLLER");
+                    dynamicVotersBuilder.append(prefix);
+                    prefix = ",";
+                    dynamicVotersBuilder.append(String.format("%d@localhost:%d:%s",
```

Comment:
Given that conflicts caused by base64 in converting UUID to String, maybe we can create `DynamicVoter` directly to avoid possible bugs in the future. for example:
```java
                List<DynamicVoter> voters = new ArrayList<>();
                for (TestKitNode controllerNode : nodes.controllerNodes().values()) {
                    int port = socketFactoryManager.
                        getOrCreatePortForListener(controllerNode.id(), "CONTROLLER");
                    voters.add(new DynamicVoter(controllerNode.metadataUuid(), controllerNode.id(), "localhost", port));
                }
                formatter.setInitialVoters(new DynamicVoters(voters));
```

Reply:
Well, testing the parsing of voter strings is actually a good thing, that we want the tests to do. :)